File: csir_intelligent_manipulator/csir_wam_object_manipulation/bhand_grasp_controller/test/ntest_bhand_grasp_controller.py
#! /usr/bin/env python

# Copyright (c) 2014 CSIR. All rights reserved.
# Use of this source code is governed by the MIT license that can be found in
# the LICENSE file.

""" Node tests for bhand_grasp_controller

    Classes:
        BHandGraspControllerNTest: test fixture
"""

PKG = "bhand_grasp_controller"
import roslib
roslib.load_manifest(PKG)
import rospy
import unittest
import sensor_msgs.msg
import object_manipulation_msgs.msg
import wam_srvs.srv
import actionlib
import actionlib_msgs
from threading import Thread
from bhand_grasp_controller import *

class BHandGraspControllerNTest(unittest.TestCase):
    """ Node test fixture for testing bhand_grasp_controller.

    Attributes:
        _bhand_pub: Publishes synthetic joint states of the Barrett Hand.
        _bhand_thread: Thread that publishes hand joint states.
        _bhand_names: Names of joints in Barrett Hand used in joint state
            messages.
        _bhand_pos: List of hand joint values in radians.
        _bhand_pub_freq: The rate in Hz at which the joint state is published.
    """

    @classmethod
    def setUpClass(cls):
        """ Initialise class members once per test suite.
        """
        cls._bhand_dof = BHandGraspController._DOF
        cls._bhand_names = BHandGraspController._BHAND_NAMES
        cls._bhand_pub_freq = BHandGraspController._BHAND_PUB_FREQ
        cls._bhand_pos = [0.0 for i in range(cls._bhand_dof)]
        cls._proximal_finger_indeces =\
            BHandGraspController._PROXIMAL_JOINTS
        rospy.init_node("ntest_bhand_grasp_controller") # to __main__
        cls._bhand_pub = rospy.Publisher("bhand/joint_states",
                                          sensor_msgs.msg.JointState)
        cls._grasp_action_client = actionlib.SimpleActionClient(
            "posture_action_name",
            object_manipulation_msgs.msg.GraspHandPostureExecutionAction)
        cls._grasp_action_client.wait_for_server() 
        cls._bhand_spread_srv = rospy.Service("bhand/spread_pos",
                                               wam_srvs.srv.BHandSpreadPos,
                                               cls.sim_bhand_spread)
        cls._bhand_grasp_srv = rospy.Service("bhand/grasp_pos",
                                              wam_srvs.srv.BHandGraspPos,
                                              cls.sim_bhand_grasp)
        cls._bhand_thread = Thread(target = cls.pub_bhand)
        cls._bhand_thread.start()

    def setUp(self):
        """ Initialise class members for test.
        """
        self._bhand_dof = BHandGraspController._DOF
        self._bhand_pub_freq = BHandGraspController._BHAND_PUB_FREQ
        self._grasp_goal =\
            object_manipulation_msgs.msg.GraspHandPostureExecutionGoal()
        #self._bhand_pos = [0.0 for i in range(self._bhand_dof)]
        #self._proximal_finger_indeces =\
            #BHandGraspController._PROXIMAL_JOINTS

    def tearDown(self):
        """ Release resources in preparation for next test.
        """
        #self._bhand_spread_srv.shutdown("Test complete")
        #self._bhand_grasp_srv.shutdown("Test complete")
        pass

    @classmethod
    def pub_bhand(cls):
        """ Publish the joint state in _joint_pos on topic bhand/joint_states.
        """
        rospy.logdebug("Publishing BHand")
        pub_rate = rospy.Rate(cls._bhand_pub_freq)
        while not rospy.is_shutdown():
            bhand_msg = sensor_msgs.msg.JointState()
            bhand_msg.header.stamp = rospy.Time.now()
            bhand_msg.name = cls._bhand_names
            bhand_msg.position = cls._bhand_pos
            bhand_msg.velocity = [0.0 for i in range(cls._bhand_dof)]
            bhand_msg.effort = [0.0 for i in range(cls._bhand_dof)]
            cls._bhand_pub.publish(bhand_msg)
            pub_rate.sleep()

    def query_grasp_service(self):
        """ Query the GraspStatus service.

        Returns:
            True if hand occupied, False otherwise.
        """
        rospy.logdebug("Querying grasp status")
        rospy.wait_for_service('grasp_query_name')
        try:
            grasp_query_client = rospy.ServiceProxy("grasp_query_name",
                                                    object_manipulation_msgs.srv.GraspStatus)
            res = grasp_query_client(object_manipulation_msgs.msg.Grasp())
            rospy.logdebug("Received grasp status")
            return res.is_hand_occupied
        except rospy.ServiceException, e:
            print "Service call failed: %s"%e

    def send_goal(self, goal):
        """ Send grasp goal to grasp actionserver.

        Arguments:
            goal (int): PRE_GRASP=1, GRASP=2, RELEASE=3
        """
        rospy.logdebug("Goal")
        self._grasp_goal.goal = goal
        self._grasp_action_client.send_goal(self._grasp_goal)
        self._grasp_action_client.wait_for_result()

    @classmethod
    def sim_bhand_spread(cls, bhand_spread_req):
        """ Simulate the BHandSpreadPos service in the Barrett wam_node.

        Changes the finger spread angle in the hand joint state.
        Arguments:
            bhand_spread_req (BHandSpreadPosRequest): The requested spread angle
                in radians
        """
        rospy.logdebug("Spreading")
        cls._bhand_pos[3] = bhand_spread_req.radians # 3 is the index of the spread angle
        return wam_srvs.srv.BHandSpreadPosResponse()

    @classmethod
    def sim_bhand_grasp(cls, bhand_grasp_req):
        """ Simulate the BHandGraspPos service in the Barrett wam_node.

        The three proximal joint angles of the Barrett hand are set to the
        requested angle.
        Arguments:
            bhand_grasp_req (BHandGraspPosRequest): The requested proximal joint
                angles in radians.
        """
        rospy.logdebug("Grasping")
        for i in cls._proximal_finger_indeces:
            cls._bhand_pos[i] = bhand_grasp_req.radians
        return wam_srvs.srv.BHandGraspPosResponse()

    def test_grasp_query(self):
        """ Test GraspStatus service.

        # Verify that initially open
        # Simulate close - change self._bhand_pos directly
        # Verify occupied
        # Simulate partially open
        # Verify occupied
        # Simulate open
        # Verify unoccupied
        """
        rospy.sleep(10.0 / self._bhand_pub_freq) # Wait for hand initialisation
        is_occupied = self.query_grasp_service()
        self.assertFalse(is_occupied) # Verify unoccupied
        for i in self.__class__._proximal_finger_indeces: # Sim close
            self.__class__._bhand_pos[i] = BHandGraspController._MAX_RANGE
        rospy.sleep(1.5 / self._bhand_pub_freq) # Wait for new joints to be published
        is_occupied = self.query_grasp_service()
        self.assertTrue(is_occupied) # Verify occupied
        for i in self.__class__._proximal_finger_indeces: # Sim partially open
            self.__class__._bhand_pos[i] = BHandGraspController._MAX_RANGE * 0.5
        rospy.sleep(1.5 / self._bhand_pub_freq) # Wait for new joints to be published
        is_occupied = self.query_grasp_service()
        self.assertTrue(is_occupied) # Verify occupied
        for i in self.__class__._proximal_finger_indeces: # Sim open
            self.__class__._bhand_pos[i] = 0.0
        rospy.sleep(1.5 / self._bhand_pub_freq) # Wait for new joints to be published
        is_occupied = self.query_grasp_service()
        self.assertFalse(is_occupied) # Verify unoccupied

    def test_grasp_execution(self):
        """ Test GraspHandPostureExecution action.

        Do twice
        # Send pre-grasp goal
        # Verify open
        # Send grasp goal
        # Verify closed
        # Send release goal
        # Verify open
        """
        rospy.sleep(10.0 / self._bhand_pub_freq) # Wait for hand initialisation
        for i in range(2):
            self.send_goal(
                object_manipulation_msgs.msg.GraspHandPostureExecutionGoal.PRE_GRASP)
            rospy.sleep(2.5 / self._bhand_pub_freq) # Wait for new joints to be published
            self.assertFalse(self.query_grasp_service()) # Verify open
            self.send_goal(
                object_manipulation_msgs.msg.GraspHandPostureExecutionGoal.GRASP)
            rospy.sleep(2.5 / self._bhand_pub_freq) # Wait for new joints to be published
            self.assertTrue(self.query_grasp_service()) # Verify occupied
            self.send_goal(
                object_manipulation_msgs.msg.GraspHandPostureExecutionGoal.RELEASE)
            rospy.sleep(2.5 / self._bhand_pub_freq) # Wait for new joints to be published
            self.assertFalse(self.query_grasp_service()) # Verify open

if __name__ == "__main__":
    import rostest
    rostest.rosrun(PKG, "ntest_bhand_grasp_controller",
                   BHandGraspControllerNTest)

File: csir_intelligent_manipulator/csir_wam_object_manipulation/bhand_grasp_controller/test/ntest_bhand_grasp_controller.cpp
/*
 * Copyright (c) 2014 csir_intelligent_manipulator authors. All rights reserved.
 * Use of this source code is governed by the MIT license that can be found in
 * the LICENSE file.
 * Author: Jimmy Kizito
 */

#include "bhand_grasp_controller/bhand_grasp_controller.h"
#include <sensor_msgs/JointState.h>
#include <actionlib/client/simple_action_client.h>
#include <actionlib/client/terminal_state.h>
#include <object_manipulation_msgs/GraspHandPostureExecutionAction.h>
#include <wam_srvs/BHandSpreadPos.h>
#include <wam_srvs/BHandGraspPos.h>
#include <boost/thread.hpp>
#include <vector>
#include <ros/ros.h>
#include <string>
#include <gtest/gtest.h>

/*!
 * @brief Node test fixture for testing bhand_grasp_controller.
 */
class BHandGraspControllerNTest : public ::testing::Test
{
 public:
  /*!
   * @brief Initialise class members before test
   */
  BHandGraspControllerNTest() :
      bHandPos_(BHandGraspController::DOF_, 0.0),
      bHandNames_(BHandGraspController::BHAND_NAMES_),
      spinner_(1),
      graspActionClient_("posture_action_name", true)
  {
    bool hasActionServerStarted =
        graspActionClient_.waitForServer(ros::Duration(10.0));
    EXPECT_TRUE(hasActionServerStarted) << "Action server did not start";
    bHandPub_ = node_.advertise<sensor_msgs::JointState>("bhand/joint_states", 1);
    bHandSpreadServiceServer_ =
        node_.advertiseService("bhand/spread_pos",
                               &BHandGraspControllerNTest::simBHandSpread,
                               this);
    bHandGraspServiceServer_ =
        node_.advertiseService("bhand/grasp_pos",
                               &BHandGraspControllerNTest::simBHandGrasp,
                               this);
    graspQueryClient_ =
        node_.serviceClient<object_manipulation_msgs::GraspStatus>(
            "grasp_query_name");
    spinner_.start();
    pubBHandThread_ = boost::thread(&BHandGraspControllerNTest::pubBHand, this);
  }
  /*!
   * @brief Stop thread before destroying
   */
  virtual ~BHandGraspControllerNTest()
  {
    pubBHandThread_.interrupt();
    pubBHandThread_.join();
  }
  /*!
   * @brief Puclish the joint state in bHandPos_ on topic bhand/joint_states.
   */
  void pubBHand()
  {
    ros::Rate pubRate(BHandGraspController::PUB_FREQ_);
    while (ros::ok())
    {
      try
      {
        boost::this_thread::interruption_point();
        sensor_msgs::JointState bHandMsg;
        bHandMsg.header.stamp = ros::Time::now();
        bHandMsg.name = bHandNames_;
        bHandMsg.position = bHandPos_;
        bHandMsg.velocity = std::vector<double>(BHandGraspController::DOF_, 0.0);
        bHandMsg.effort = std::vector<double>(BHandGraspController::DOF_, 0.0);
        bHandPub_.publish(bHandMsg);
        pubRate.sleep();
      }
      catch (boost::thread_interrupted &e)
      {
        break;
      }
    }
  }
  /*!
   * @brief Query the GraspStatus service.
   * @return true if hand occupied, false otherwise
   */
  bool queryGraspService()
  {
    ROS_DEBUG("Querying grasp status");
    object_manipulation_msgs::GraspStatus serviceRequest;
    if (graspQueryClient_.call(serviceRequest))
    {
      ROS_DEBUG("Received grasp status");
      return serviceRequest.response.is_hand_occupied;
    }
    else
    {
      //can only use assertions in void returning functions
      EXPECT_TRUE(false) << "Failed to call GraspStatus service";
      return false;
    }
  }
  /*!
   * @brief Send grasp goal to grasp action server.
   * @param [in] goal PRE_GRASP=1, GRASP=2, RELEASE=3
   */
  void sendGoal(const int &goal)
  {
    graspGoal_.goal = goal;
    graspActionClient_.sendGoal(graspGoal_);
    graspActionClient_.waitForResult();
  }
  /*!
   * @brief Simulate the BHandSpreadPos service in the Barrett wam_node.
   * @details Changes the finger spread angle in the hand joint state.
   * @param [in] req The requested spread angle in radians.
   * @param [out] res Unused
   */
  bool simBHandSpread(wam_srvs::BHandSpreadPos::Request &req,
                      wam_srvs::BHandSpreadPos::Response &res)
  {
    ROS_DEBUG("Spreading");
    bHandPos_.at(3) = req.radians;
    return true;
  }
  /*!
   * @brief Simulate the BHandGraspPos service in the Barrett wam_node.
   * @details The three proximal joints of the Barrett hand are set to the
   * requsted angle.
   * @param [in] req The requested proximal joint angles in radians.
   * @param [out] res Unused
   */
  bool simBHandGrasp(wam_srvs::BHandSpreadPos::Request &req,
                     wam_srvs::BHandSpreadPos::Response &res)
  {
    ROS_DEBUG("Grasping");
    for (std::vector<int>::const_iterator it =
         BHandGraspController::PROXIMAL_JOINTS_.begin(); it !=
         BHandGraspController::PROXIMAL_JOINTS_.end(); ++it)
    {
      bHandPos_.at(*it) = req.radians;
    }
    return true;
  }
  /*!
   * @brief Sleep for a fracional number of hand publisher periods rounded to
   * nearest millisecond.
   * @param [in] numPeriods Number of periods to sleep for.
   */
  void sleepFor(const double &numPeriods)
  {
    boost::this_thread::sleep(
        boost::posix_time::milliseconds(numPeriods /
                                        BHandGraspController::PUB_FREQ_*1000));
  }
  /*! Listo of hand joint values in radians*/
  std::vector<double> bHandPos_;

 protected:
  virtual void SetUp()
  {
  }

  virtual void TearDown()
  {
  }

 private:
  /*! Names of joints in Barrett Hand used in joint state messages*/
  std::vector<std::string> bHandNames_;
  ros::AsyncSpinner spinner_; // process callbacks in separate thread
  boost::thread pubBHandThread_;
  ros::NodeHandle node_;
  object_manipulation_msgs::GraspHandPostureExecutionGoal graspGoal_;
  ros::Publisher bHandPub_;
  actionlib::SimpleActionClient<object_manipulation_msgs::GraspHandPostureExecutionAction> graspActionClient_;
  ros::ServiceServer bHandSpreadServiceServer_;
  ros::ServiceServer bHandGraspServiceServer_;
  ros::ServiceClient graspQueryClient_;
};

/*!
 * @brief Test GraspStatus service.
 * @details Sleep used to allow for new joint state to be published.
 * # Wait for hand to initialise
 * # Verify that hand initially open
 * # Simulate hand closed
 * # Verify hand occupied
 * # Simulate hand partially open
 * # Verify hand occupied
 * # Simulte hand open
 * # Verify hand unoccupied
 */
TEST_F(BHandGraspControllerNTest, graspQuery)
{
  sleepFor(10.0);
  ROS_DEBUG("Test");
  bool isOccupied = queryGraspService();
  ASSERT_FALSE(isOccupied) << "Hand not initially empty";
  std::vector<int>::const_iterator it; 
  for (it = BHandGraspController::PROXIMAL_JOINTS_.begin(); it !=
       BHandGraspController::PROXIMAL_JOINTS_.end(); ++it)
  { // sim close
    bHandPos_.at(*it) = BHandGraspController::MAX_RANGE_;
  }
  sleepFor(1.5);
  isOccupied = queryGraspService();
  ASSERT_TRUE(isOccupied) << "Hand should be occupied";
  for (it = BHandGraspController::PROXIMAL_JOINTS_.begin(); it !=
       BHandGraspController::PROXIMAL_JOINTS_.end(); ++it)
  { // sim partially open
    bHandPos_.at(*it) = BHandGraspController::MAX_RANGE_ * 0.5;
  }
  sleepFor(1.5);
  isOccupied = queryGraspService();
  ASSERT_TRUE(isOccupied) << "Hand should be occupied";
  for (it = BHandGraspController::PROXIMAL_JOINTS_.begin(); it !=
       BHandGraspController::PROXIMAL_JOINTS_.end(); ++it)
  { // sim open 
    bHandPos_.at(*it) = 0.0;
  }
  sleepFor(1.5);
  isOccupied = queryGraspService();
  ASSERT_FALSE(isOccupied) << "Hand should be empty";
}

/*!
 * @brief Test GraspHandPostureExecution action.
 * @details Do twice:
 * # Send pre-grasp goal
 * # Verify hand open
 * # Send grasp goal
 * # Verify hand closed
 * # Send release goal
 * # Verify hand open
 */
TEST_F(BHandGraspControllerNTest, graspExecution)
{
  //boost::thread bHandThread(BHandGraspControllerNTest::pubBHand, this);
  for (int i = 0; i < 2; ++i)
  {
    sendGoal(object_manipulation_msgs::GraspHandPostureExecutionGoal::PRE_GRASP);
    sleepFor(2.5);
    ASSERT_FALSE(queryGraspService()) << "Hand should be empty";
    sendGoal(object_manipulation_msgs::GraspHandPostureExecutionGoal::GRASP);
    sleepFor(2.5);
    ASSERT_TRUE(queryGraspService()) << "Hand should be occupied";
    sendGoal(object_manipulation_msgs::GraspHandPostureExecutionGoal::RELEASE);
    sleepFor(2.5);
    ASSERT_FALSE(queryGraspService()) << "Hand should be empty";
  }
}

int main(int argc, char **argv)
{
  ::testing::InitGoogleTest(&argc, argv);
  ros::init(argc, argv, "ntest_bhand_grasp_controller");
  return RUN_ALL_TESTS();
}

